include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3380,6 +3381,194 @@ namespace internal
+    static_assert(dim >= 0 && dim <= 3, "Only dim=0,1,2,3 implemented");
+
+    // we only need the value on faces of a 1d element
+    if (dim == 0)
+      {
+        return values[0];
+      }
+
+    using Number3 = typename ProductTypeNoPoint<Number, Number2>::type;
+
+    // Go through the tensor product of shape functions and interpolate
+    // with optimal algorithm
+    Number3 result = {};
+    for (int i2 = 0, i = 0; i2 < (dim > 2 ? n_shapes : 1); ++i2)
+      {
+        Number3 inner_result;
+        // Generate separate code with known loop bounds for the most common
+        // cases
+        if (n_shapes == 2)
+          inner_result = do_interpolate_xy_value<dim, 2, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        else if (n_shapes == 3)
+          inner_result = do_interpolate_xy_value<dim, 3, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        else if (n_shapes == 4)
+          inner_result = do_interpolate_xy_value<dim, 4, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        else if (n_shapes == 5)
+          inner_result = do_interpolate_xy_value<dim, 5, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        else if (n_shapes == 6)
+          inner_result = do_interpolate_xy_value<dim, 6, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        else
+          inner_result = do_interpolate_xy_value<dim, -1, Number2, Number>(
+            values, renumber, shapes, n_shapes, i);
+        if (dim == 3)
+          {
+            // Interpolation + derivative in z direction
+            result += inner_result * shapes[i2][0][2];
+          }
+        else
+          {
+            result = inner_result;
+          }
+      }
+
+    return result;
+  }
+
+
+
+  template <int dim, typename Number, typename Number2>
+  inline typename ProductTypeNoPoint<Number, Number2>::type
+  evaluate_tensor_product_value_linear(
+    const unsigned int               n_shapes,
+    const Number *                   values,
+    const Point<dim, Number2> &      p,
+    const std::vector<unsigned int> &renumber = {})
+  {
+    (void)n_shapes;
+    static_assert(dim >= 0 && dim <= 3, "Only dim=0,1,2,3 implemented");
+
+    using Number3 = typename ProductTypeNoPoint<Number, Number2>::type;
+
+    AssertDimension(n_shapes, 2);
+    for (unsigned int i = 0; i < renumber.size(); ++i)
+      AssertDimension(renumber[i], i);
+
+    if (dim == 0)
+      {
+        // we only need the value on faces of a 1d element
+        return values[0];
+      }
+    else if (dim == 1)
+      {
+        return (1. - p[0]) * values[0] + p[0] * values[1];
+      }
+    else if (dim == 2)
+      {
+        const Number2 x0 = 1. - p[0], x1 = p[0];
+        const Number3 tmp0   = x0 * values[0] + x1 * values[1];
+        const Number3 tmp1   = x0 * values[2] + x1 * values[3];
+        const Number3 mapped = (1. - p[1]) * tmp0 + p[1] * tmp1;
```

Comment:
How about doing the operations the other way around, transforming 2 subtract, 3 mult, 3 FMAs into 3 subtract + 3 FMAs:
```suggestion
        const Number3 tmp0   = values[0] + p[0] * (values[1] - values[0]);
        const Number3 tmp1   = values[2] + p[0] * (values[3] - values[2]);
        const Number3 mapped = tmp0 + p[1] * (tmp1 - tmp0);
```
and the same in the 3D version below? In 3D, we would go from 3 subtract, 7 mult, 7 FMAs to 7 subtract + 7 FMAs. Even though an FP addition has similar costs as a multiply, it should be cheaper. And we overall reduce the number of FP operations by `dim`. Given enough out-of-order-execution resources, this should have an overall positive effect, even though the length of the critical path will not change: Before, we would have the critical path as `subtract for x0` -> `FMA tmp0/1` -> `mult with tmp0` -> `FMA for mapped` (other operations running in parallel), whereas we now have `subtract values[1]-values[0]` -> `FMA tmp0/2` -> `subtract tmp1-tmp0` -> `FMA for mapped`. An interesting detail is the question whether the possibly different types, `scalar` for `values[i]` and vectorized for `p[0]`, might lead to observable slowdowns because broadcast `memory -> vector register` (as when loading `values[0]` and consuming it in an operation with `x0`) versus first performing the scalar subtraction and then broadcasting `register -> vector register` leads to visible performance differences. The latter has lower throughput on Intel architectures but not on AMD.

Reply:
I do not see significantly better timings for this on our Intel server CPU. Can you test this on an AMD machine? Here is the branch with the alternative implementation https://github.com/bergbauer/dealii/tree/evaluate_integrate_values_test

Reply:
But the timings are also not worse on Intel? To be clear, the problem on Intel is that `bcast registor -> vector register` goes to the same execution port as FMAs, so we actually do not gain much. But for the derivative we do. Let us keep this change out of the present PR and address it later.

Reply:
The difference in timings is in the noise. I agree, let's address this in a follow-up!